fil_fungible_token/src/receiver/mod.rs

```diff
@@ -1 +1,56 @@
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::{address::Address, econ::TokenAmount, error::ExitCode};
+use num_traits::Zero;
+use types::TokensReceivedParams;
+
+use crate::runtime::messaging::{Messaging, RECEIVER_HOOK_METHOD_NUM};
+use crate::token::TokenError;
+
 pub mod types;
+
+#[derive(Debug)]
+pub struct ReceiverHookGuard<T> {
+    address: Address,
+    params: Option<TokensReceivedParams>,
+    return_value: Option<T>,
+}
+
+impl<T> ReceiverHookGuard<T> {
+    pub fn new(address: Address, params: TokensReceivedParams, return_value: T) -> Self {
+        ReceiverHookGuard { address, params: Some(params), return_value: Some(return_value) }
+    }
+    pub fn call(&mut self, msg: &dyn Messaging) -> std::result::Result<T, TokenError> {
+        if self.params.is_none() {
+            return Err(TokenError::ReceiverHookGuardAlreadyCalled);
+        }
+
+        // this will leave self.params set to None, so a further attempt to call() will fail
+        let params = self.params.take().unwrap();
```

Comment:
I find this use of an option to be unnecessarily obscure, vs an explicit "was_called" field. I don't think we're saving any memory or anything vs that approach.

Reply:
Previous iterations required it because of ownership and lifetime concerns. I don't think there's anything special in the `TokensReceivedParams` that will stand in the way now, so I'll switch it over.